### PR TITLE
fix #10248

### LINKED ISSUE #10248: 使用dataset数据时，series.label.formatter使用function(params){}，params.value的值为dataset中所有列，而不是该series对应的那一列的数据
### Version
4.2.1

### Steps to reproduce
https://echarts.baidu.com/examples/editor.html?c=dataset-simple1

option = {
    legend: {},
    tooltip: {},
    dataset: {
        dimensions: ['product', '2015', '2016', '2017'],
        source: [
            {product: 'Matcha Latte', '2015': 43.3, '2016': 85.8, '2017': 93.7},
            {product: 'Milk Tea', '2015': 83.1, '2016': 73.4, '2017': 55.1},
            {product: 'Cheese Cocoa', '2015': 86.4, '2016': 65.2, '2017': 82.5},
            {product: 'Walnut Brownie', '2015': 72.4, '2016': 53.9, '2017': 39.1}
        ]
    },
    xAxis: {type: 'category'},
    yAxis: {},
    // Declare several bar series, each will be mapped
    // to a column of dataset.source by default.
    series: [
        {type: 'bar',
        encode: {x: 'product', y:'2015'},
            label: {
                show: true,
                formatter: a => JSON.stringify(a.value)
            }
        }
    ]
};

猜测问题出在https://github.com/apache/incubator-echarts/blob/master/src/chart/bar/helper.js#L29
参数中缺一个dimIndex的参数，其它图表类型下也存在类似问题，不仅仅是bar
最终参数传递到
https://github.com/apache/incubator-echarts/blob/master/src/model/mixin/dataFormat.js#L87

### What is expected?
series.label.formatter = function(params){}
params.value应该是该series对应的列的值

### What is actually happening?
params.value获得的是所有列的值的数组

### PR DESCRIPTION
**Pull request by quick-sort**

### CODE REVIEW

File: src/model/mixin/dataFormat.js [hunk 1]
```diff
@@ -94,6 +94,12 @@ export default {
         );
 
         if (typeof formatter === 'function') {
+            if ((params.value instanceof Array) && dimIndex == null) {
+              let labelDims = data.mapDimension('defaultedLabel', true)
```
Comment (Ovilia) [inline src/model/mixin/dataFormat.js]: Please use `var` instead of `let`.

File: src/model/mixin/dataFormat.js [hunk 2]
```diff
@@ -94,6 +94,12 @@ export default {
...
+              if (labelDims instanceof Array) {
+                params.value = labelDims.map(i => retrieveRawValue(data, dataIndex, i))
```
Comment (Ovilia) [inline src/model/mixin/dataFormat.js]: Arrow function is not supported.

File: src/model/mixin/dataFormat.js [hunk 3]
```diff
@@ -94,6 +94,12 @@ export default {
         );
 
         if (typeof formatter === 'function') {
+            if ((params.value instanceof Array) && dimIndex == null) {
```
Comment (Ovilia) [inline src/model/mixin/dataFormat.js]: Please use `zrUtil.isArray` to check Array. Search for usage.

File: src/model/mixin/dataFormat.js (hunk 1 above)
Comment (Ovilia) [inline src/model/mixin/dataFormat.js]: Please add `;` at the end.

**Comment by quick-sort** [comment]
changes resolved

**Comment by 100pah** [comment]
Thanks @quick-sort .

 The `defaultedLabel` is by "guess". So I think we should not restrict that only give the values of the `defaultedLabel` to users, but give all of the possible values. 
But yes, in the `param` we'd better provide the mapping info (x -> which dimension, y -> which dimension) to users, by which the corresponding values can be fetched in each series.

**Pull request merged**